### ISSUE TITLE
Performance issues in big files

### ISSUE THREAD

**Issue by bbatsov** [Malabarba/aggressive-indent-mode#73]
Try running aggressive-indent on https://github.com/bbatsov/zenburn-emacs/blob/master/zenburn-theme.el

Every edit gets super slow.

**Comment by Malabarba** [comment]
Probably a duplicate of #58. The relevant parameter is not the size of the file, but the size of the function. 
For technical reasons, when we're reindenting, we need to go at least until the start of the next function, which means we need to reindent until the end of the current function.

Aggressive-indent tries to be polite in those cases. It uses `abort-on-input` so that you can still type while it's working in the background. But that's not perfect. :-(

**Comment by bbatsov** [comment]
If function size is the issue, perhaps it should be auto-disabled if a function is bigger than some number of characters. Sprinkling buffer local vars all over the code is not exactly fun.

**Comment by Malabarba** [comment]
Yes, something like that has been on my todo list for a while now.

Ideally, we'd actually detect cases where indentation is very slow and disable the mode or switch to a less aggressive method.
In practice, just counting function size might be more realistic.

**Comment by BrunoBonacci** [comment]
Hi, I'm running this issue as well, in my case functions are very small but file is quite big (2000+ lines).
I know that 2k lines of source code in a single file isn't a "normal" case, but I have a large clojure file with long comments which I use for introductory course to Clojure and several EDN files which are quite big.
Working with them while the aggressive indent mode is enabled is impossible.
It would be nice to disable automatically the mode when the buffer is more than 1000 lines.

**Comment by expez** [comment]
This is killing me when working with liberator.  The resource functions are really one function, consisting of many smaller functions, and these can span hundreds of lines in the code base at work.

**Comment by Malabarba** [comment]
I'll see if I can find a solution for this over the week.

**Comment by Malabarba** [comment]
Ok, I've added a variable that fine-tunes the delay used before starting to indent. The name is `aggressive-indent-sit-for-time`. Try increasing or decreasing it and see if it helps. While aggressive-indent is sitting (and while it's indenting) the code is wrapped in a `while-no-input`. So Emacs **really** shouldn't hang in the meantime.

**Comment by seagle0128** [comment]
Same issue here with the large files. Very slow to save...

**Comment by Salompas** [comment]
I am also having a very very big slow down in Emacs when editing Matlab scripts that are small (150 lines, attached below). Just by typing Emacs starts "lagging" and inputting text gets jittery.

[separateResultsByMoneyness.zip](https://github.com/Malabarba/aggressive-indent-mode/files/1325485/separateResultsByMoneyness.zip)

.

**Comment by seagle0128** [comment]
@Salompas I have no performance issue while editing the file you provided. I am working with GNU Emacs 25.3 on macOS 10.13.
